src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+        //    if (this.State == CommunicationState.Opened)
+        //    {
+        //        if (this.SecurityProtocolFactory != null)
+        //        {
+        //            result = this.SecurityProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //        else if (this.SessionServerSettings != null && this.SessionServerSettings.SessionProtocolFactory != null)
+        //        {
+        //            result = this.SessionServerSettings.SessionProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //    }
+        //    if (result == null)
+        //    {
+        //        result = base.GetProperty<EndpointIdentity>();
+        //    }
+        //    this.identity = result;
+        //}
+
+
+
+        void ThrowIfProtocolFactoryNotSet()
+        {
+            if (this.securityProtocolFactory == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.SecurityProtocolFactoryShouldBeSetBeforeThisOperation)));
+            }
+        }
+        public async Task<IServiceChannelDispatcher> CreateServiceChannelDispatcherAsync(IChannel outerChannel)
+        {
+            this.outerChannel = outerChannel;
+            IChannel securityChannel = GetInnerChannel(outerChannel);
+            var sessionIdleManager = securityChannel.GetProperty<ServiceChannel.SessionIdleManager>();
+            IChannelBinder binder = null;
+            if (securityChannel is IReplyChannel)
+            {
+                var rcbinder = securityChannel.GetProperty<SecurityReplyChannelBinder>();
+                rcbinder.Init(securityChannel as IReplyChannel, BaseAddress);
+                binder = rcbinder;
+            }
+            else
+            {
+                throw new NotImplementedException();
+            }
+            //Open the session server 
+            await this.SessionServerSettings.OnOpenAsync(ServiceDefaults.OpenTimeout);
+
+            // TODO: MUST add the proper messsage
+            if (this.SecurityAuthChannelDispatcher == null)
+                throw new Exception("SecurityAuthChannelDispatcher can't be null");
+
+            ServiceDispatcher securitySessionAuthDispatcher = new ServiceDispatcher(this.SecurityAuthChannelDispatcher);
+            var channelHandler = new ChannelHandler(Binding.MessageVersion, binder, securityChannel.GetProperty<ServiceThrottle>(),
+            securitySessionAuthDispatcher, /*wasChannelThrottled*/ false, sessionIdleManager);
+            var channelDispatcher = channelHandler.GetDispatcher();
+            outerChannel.ChannelDispatcher = securityChannel.ChannelDispatcher = channelDispatcher;
+            await channelHandler.OpenAsync();
+            return channelDispatcher;
+        }
+
+        private IChannel GetInnerChannel(IChannel outerChannel)
+        {
+            IChannel securityChannel = null;
+            /* TODO once we add more features
```

Comment:
You should have code similar to this at the start:
```c#
                SecurityProtocol securityProtocol = SecurityProtocolFactory.CreateSecurityProtocol(null, null, this.listenerSecurityProtocolState,
                    typeof(TChannel) == typeof(IReplyChannel) || typeof(TChannel) == typeof(IReplySessionChannel), TimeSpan.Zero);
```
Maybe without the listenerSecurityProtocolState as I couldn't find anywhere where that was set to anything other than null.
Then you should pass the securityProtocol and settingsLifetimeManager to the channel's constructor. The channel needs to use that, not the listener.